### PR TITLE
util + prov/rxm: Serialize recv list access and read multiple completions when progressing.

### PR DESCRIPTION
**Pull request by a-ilango**
-	util: Add a lock to util_ep
-	prov/rxm: Serialize access to recv_list and unexp_msg_list
-	prov/rxm: Read multiple completions per progress

### CODE REVIEW

File: prov/rxm/src/rxm_cq.c [hunk 1]
```diff
@@ -512,20 +514,25 @@ static ssize_t rxm_cq_read(struct fid_cq *msg_cq, struct fi_cq_msg_entry *comp)
...
-			goto err;
-	} while (ret > 0);
+		comp_read += ret;
+		if (ret) {
+			ret = rxm_cq_handle_comp(rxm_ep, &comp);
+			if (ret)
+				goto err;
+		}
+	} while ((comp_read < (MIN(rxm_ep->msg_info->tx_attr->size,
+			      rxm_ep->msg_info->rx_attr->size) / 2)) &&
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: I would save this MIN value, so the progress call doesn't end up calculating it every time the call is made (and per iteration).

File: prov/rxm/src/rxm_cq.c [hunk 2]
```diff
@@ -512,20 +514,25 @@ static ssize_t rxm_cq_read(struct fid_cq *msg_cq, struct fi_cq_msg_entry *comp)
...
-		ret = rxm_cq_read(msg_cq, &comp);
+		ret = rxm_cq_read(rxm_ep->msg_cq, &comp);
 		if (ret < 0)
 			goto err;
 
-		ret = rxm_cq_handle_comp(&comp);
-		if (ret)
-			goto err;
-	} while (ret > 0);
+		comp_read += ret;
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: No need to add if ret == 0, and there's a check below for positive ret.

**Comment by a-ilango** [comment]
Thanks. Updated the PR with desired changes.

File: prov/rxm/src/rxm_cq.c [hunk 3]
```diff
@@ -512,20 +514,23 @@ static ssize_t rxm_cq_read(struct fid_cq *msg_cq, struct fi_cq_msg_entry *comp)
...
-		if (ret)
-			goto err;
-	} while (ret > 0);
+		if (ret) {
+			comp_read += ret;
+			ret = rxm_cq_handle_comp(rxm_ep, &comp);
+			if (ret)
+				goto err;
+		}
+	} while (comp_read > 0 && comp_read < rxm_ep->comp_per_progress);
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: If you read one completion (comp_read = 1), then no more are available, does this loop spin?  I'm wondering if the comp_read > 0 check is what's wanted, or if it should be ret > 0

Comment (a-ilango) [inline prov/rxm/src/rxm_cq.c]: When there is just one completion available, we would get EAGAIN on the next read and return. Perhaps that's why it doesn't busy loop here. But yeah, the condition should be ret > 0 to be correct. We may need a new variable here as ret is being overwritten by call to rxm_cq_handle_comp.

Comment (a-ilango) [inline prov/rxm/src/rxm_cq.c]: How about breaking out of the loop if ret == -FI_EAGAIN and just checking if comp_read < comp_per_progress as loop condition?

**Comment by shefty** [comment]
That seems fine.  It's the comp_read > 0 check that's confusing me.

**Comment by a-ilango** [comment]
@shefty This is ready for merge if you don't have any more comments.

**Pull request merged**